lib/buildImage.nix

```diff
@@ -0,0 +1,133 @@
+  # a CI to run Nix builds.
+  initializeNixDatabase ? false
+, # If initializeNixDatabase is set to true, the uid/gid of /nix can be
+  # controlled using nixUid/nixGid.
+  nixUid ? 0
+, nixGid ? 0
+, # Deprecated: will be removed
+  contents ? null
+, meta ? { }
+,
+}:
+let
+  configFile = writeText "config.json" (builtins.toJSON config);
+  copyToRootList =
+    let derivations = if !isNull contents then contents else copyToRoot;
+    in if isNull derivations
+    then [ ]
+    else if !builtins.isList derivations
+    then [ derivations ]
+    else derivations;
+
+  # Expand the given list of layers to include all their transitive layer dependencies.
+  layersWithNested = layers:
+    let layerWithNested = layer: [ layer ] ++ (builtins.concatMap layerWithNested (layer.layers or [ ]));
+    in builtins.concatMap layerWithNested layers;
+  explodedLayers = layersWithNested layers;
+  ignore = [ configFile ] ++ explodedLayers;
+
+  closureGraphForAllLayers = closureGraph ([ configFile ] ++ copyToRootList ++ layers) ignore;
+  nixDatabase = makeNixDatabase closureGraphForAllLayers;
+  # This layer contains all config dependencies. We ignore the
+  # configFile because it is already part of the image, as a
+  # specific blob.
+
+  perms' = perms ++ lib.optionals initializeNixDatabase
+    [
+      {
+        path = nixDatabase;
+        regex = ".*";
+        mode = "0755";
+        uid = nixUid;
+        gid = nixGid;
+      }
+    ];
+
+  customizationLayer = buildLayer {
+    inherit maxLayers;
+    perms = perms';
+    copyToRoot =
+      if initializeNixDatabase
+      then copyToRootList ++ [ nixDatabase ]
+      else copyToRootList;
+    deps = [ configFile ];
+    ignore = configFile;
+    layers = layers;
+  };
+  fromImageFlag = lib.optionalString (fromImage != "") "--from-image ${fromImage}";
+  layerPaths = lib.concatMapStringsSep " " (l: l + "/layers.json") (layers ++ [ customizationLayer ]);
+  image =
+    let
+      imageName = lib.toLower name;
+      imageTag =
+        if tag != null
+        then tag
+        else
+          lib.head (lib.strings.splitString "-" (baseNameOf image.outPath));
+    in
+    runCommand "image-${baseNameOf name}.json"
+      {
+        inherit imageName meta;
+        nativeBuildInputs = [ nix2container-bin ];
+        passthru = {
+          inherit fromImage imageTag;
+          # provide a cheap to evaluate image reference for use with external tools like docker
+          # DO NOT use as an input to other derivations, as there is no guarantee that the image
+          # reference will exist in the store.
+          imageRefUnsafe = builtins.unsafeDiscardStringContext "${imageName}:${imageTag}";
+          copyToDockerDaemon = selfBuildHost.copyToDockerDaemon image;
+          copyToRegistry = selfBuildHost.copyToRegistry image;
+          copyToPodman = selfBuildHost.copyToPodman image;
+          copyTo = selfBuildHost.copyTo image;
```

Comment:
We want these to run on the build platform. These need to be explicitly picked from the `buildHost` variant because it's not possible to automatically infer which variant we want here.